src/foam/util/concurrent/AsyncAssemblyLine.java

```diff
@@ -28,27 +28,39 @@ public AsyncAssemblyLine(X x) {
  public void enqueue(Assembly job) {
    final Assembly previous;
 
-   synchronized ( startLock_ ) {
-     previous = q_;
-     q_ = job;
-     job.startJob();
-   }
-
-   pool_.submit(x_, new ContextAgent() { public void execute(X x) {
-     job.executeJob();
-
-     if ( previous != null ) previous.waitToComplete();
-
-     synchronized ( endLock_ ) {
-       job.endJob();
-       job.complete();
+   try {
+     synchronized ( startLock_ ) {
+       previous = q_;
+       q_ = job;
+       job.executeUnderLock();
```

Comment:
@kgrgreer executeUnderLock previously wasn't in Async.  Was that intentional? 
Also looking at this now, if we try/catch here it can occur inside the sync and just set q_ = null on catch.

Reply:
Forgot to add.